chrome/content/zotero/xpcom/feedReader.js

```diff
@@ -0,0 +1,527 @@
+			if (issn) info.ISSN = issn;
+			
+			let isbn = getFeedField(feed, 'prism', 'isbn')
+				|| getFeedField(feed, null, 'isbn')
+			if (isbn) info.ISBN = isbn;
+			
+			let language = getFeedField(feed, 'dc', 'language')
+				|| getFeedField(feed, null, 'language');
+			if (language) info.language = language;
+			
+			let ttl = getFeedField(feed, null, 'ttl');
+			if (ttl) info.ttl = ttl;
+			
+			return info;
+		});
+		
+		// Array of deferred item promises
+		this._feedItems = [Zotero.Promise.defer()];
+		
+		// Process items once they're available and push them into the array
+		Zotero.Promise.join(
+			this._feed.promise,
+			this._feedProperties,
+			(feed, feedInfo) => {
+				let items = feed.items;
+				if (items && items.length) {
+					for (let i=0; i<items.length; i++) {
+						let item = items.queryElementAt(i, Components.interfaces.nsIFeedEntry);
+						if (!item) continue;
+						
+						let feedItem = getFeedItem(item, feedInfo);
+						if (!feedItem) continue;
+						
+						let lastItem = this._feedItems[this._feedItems.length - 1];
+						this._feedItems.push(Zotero.Promise.defer()); // Push a new deferred promise so an iterator has something to return
+						lastItem.resolve(feedItem);
+					}
+				}
+			}
+		)
+		.finally(() => {
+			// Make sure the last promise gets resolved to null
+			let lastItem = this._feedItems[this._feedItems.length - 1];
+			lastItem.resolve(null);
+		});
+		
+		// Set up asynchronous feed processor
+		let feedProcessor = Components.classes["@mozilla.org/feed-processor;1"]
+			.createInstance(Components.interfaces.nsIFeedProcessor);
+		
+		let feedUrl = ios.newURI(url, null, null);
+		feedProcessor.parseAsync(null, feedUrl);
+		
+		feedProcessor.listener = {
+			/*
+			 * MDN suggests that we could use nsIFeedProgressListener to handle the feed
+			 * as it gets loaded, but this is actually not implemented (as of 32.0.3),
+			 * so we have to load the whole feed and handle it in handleResult.
+			 */
+			handleResult: (result) => {
+				if (!result.doc) {
+					this.terminate("No Feed");
+					return;
+				}
+				
+				let newFeed = result.doc.QueryInterface(Components.interfaces.nsIFeed);
+				this._feed.resolve(newFeed);
+			}
+		};
+		
+		Zotero.debug("FeedReader: Fetching feed from " + feedUrl.spec);
+		
+		this._channel = ios.newChannelFromURI(feedUrl);
+		this._channel.asyncOpen(feedProcessor, null); // Sends an HTTP request
+	}
+	
+	Zotero.defineProperty(FeedReader.prototype, 'feedProperties', {
+		get: function() this._feedProperties
+	});
+	
+	/*
+	 * Feed item iterator
+	 * Each iteration returns a _promise_ for an item. The promise _MUST_ be
+	 * resolved before requesting the next item.
+	 * The last item will always be resolved to `null`, unless the feed processing
+	 * is terminated ahead of time, in which case it will be rejected with the reason
+	 * for termination.
+	 */
+	Zotero.defineProperty(FeedReader.prototype, 'itemIterator', {
```

Comment:
I'm reconsidering this. I think it would make most sense to make usage `var iterator = new fr.ItemIterator()` rather than `var iterator = fr.itemIterator` since `fr.itemIterator !== fr.itemIterator` given this code. Yes, no, maybe?